### ISSUE TITLE
[Issue]: Different rendering (or error, if HiDiffusion ON) result at first generate attempt

### ISSUE THREAD

**Issue by mart-hill** [vladmandic/automatic#3198]
### Issue Description

When I change ANYTHING on the prompt, first generating attempt produces a different, 'dreamy' sort of picture, unexpected for the output I'm expecting, but the generation retry generates the different image. 

And sometimes, after loading a fresh/other SD1.5 model, then Generate (AND HiDiffusion ON) the error occurs (I've included it in the relevant part of the form). Second attempt (again, without changing anything) generates the picture correctly.

Both cases are without a refiner. I'll include an example to replicate, with SD1.5 model [revAnimated_v2-Rebirth [0ca0c297a4]](https://civitai.com/models/7371?modelVersionId=425083):

First attempt:
![00033-20240604200257_3881769075_a0bedb42-grid](https://github.com/vladmandic/automatic/assets/55754730/8689d068-0cfb-4357-8e77-a365c3514dcd)

Second attempt, the same parameters:
![00034-20240604200403_3881769075_877e17fa-grid](https://github.com/vladmandic/automatic/assets/55754730/db9bddf5-d2de-4c12-abdd-893f592665dd)

The grids should have all te parameter to reproduce the image. 🙂 

SDXL try gave me an error, when I wanted to use the 'generic' model AND the slightly modified 'generic' refiner - the UI desperately wanted to download the refiner from Huggingface (and didn't calculate the checksum of my refiner file nor did it load the model). But it generated an image (or rather a colored glass effect) nonetheless.🙂 

First attempt:
![00035-20240604201818_3881769075_708a0ec8-grid](https://github.com/vladmandic/automatic/assets/55754730/0cce8199-4d5d-4d4a-a850-abaca12f97ec)

Second attempt, the same parameters:
![00036-20240604201916_3881769075_3e197f46-grid](https://github.com/vladmandic/automatic/assets/55754730/f48c5d52-8ba5-4814-b1c3-5175ae76c1c2)

The XL examples are done without the refiner ~~as the UI doesn't want it~~. Second attempt to load (after loading a SD1.5 model and setting the refiner to None, then loading the 'generic' SDXL model) DID calculate the hash, loaded the refiner model, AND downloaded some 10 files.🙂 

The UI didn't render the image with HiRes ON, errored with `Exception: mat1 and mat2 shapes cannot be multiplied (154x2048 and 1280x1536)` - I'm probably not supposed to do that, right? 
The UI also didn't render the images with Second Pass still ON, but without HiRes - 
```
Exception: You have passed a list of generators of length 2, but requested an effective batch size of 1. Make sure the
                         batch size matches the length of the generators.
```

These XL issues might not be related, but I decided to put them together, as I encountered them during testing the different outputs I've got ALSO with XL generating.

### Version Platform Description

```
07:05:08-559597 INFO     Starting SD.Next
07:05:08-565598 INFO     Logger: file="X:\AI\automatic\sdnext.log" level=INFO size=2830067 mode=append
07:05:08-568097 INFO     Python 3.10.9 on Windows
07:05:09-372097 INFO     Version: app=sd.next updated=2024-06-02 hash=4acb0c88 branch=master
                         url=https://github.com/vladmandic/automatic/tree/master
07:05:10-045597 INFO     Platform: arch=AMD64 cpu=Intel64 Family 6 Model 85 Stepping 4, GenuineIntel system=Windows
                         release=Windows-10-10.0.19045-SP0 python=3.10.9
07:05:10-067102 INFO     nVidia CUDA toolkit detected: nvidia-smi present
07:05:12-129098 INFO     Torch 2.3.0+cu121
07:05:12-154597 INFO     Torch backend: nVidia CUDA 12.1 cuDNN 8801
07:05:12-158097 INFO     Torch detected GPU: NVIDIA GeForce RTX 3090 VRAM 24575 Arch (8, 6) Cores 82
07:05:21-032598 INFO     Startup: standard
07:05:21-034099 INFO     Verifying requirements
07:05:21-040097 INFO     Verifying packages
07:05:21-041596 INFO     Verifying submodules
07:05:42-919096 ERROR    Error running extension installer: X:\AI\automatic\extensions\openOutpaint-webUI-extension\install.py
07:06:27-036096 INFO     Extensions enabled: ['Lora', 'sd-extension-chainner', 'sd-extension-system-info', 'sd-webui-agent-scheduler',
                         'sdnext-modernui', 'stable-diffusion-webui-rembg', 'a1111-sd-webui-tagcomplete', 'adetailer',
                         'advanced_euler_sampler_extension', 'CFG-Schedule-for-Automatic1111-SD', 'embedding-inspector', 'model-keyword',
                         'OneButtonPrompt', 'openOutpaint-webUI-extension', 'openpose-editor', 'prompt-fusion-extension', 'PXL8',
                         'sd-dynamic-prompts', 'sd-extension-nudenet', 'sd-infinity-grid-generator-script', 'sd-model-preview-xd', 'sd-pixel',
                         'sd-webui-ar', 'sd-webui-aspect-ratio-helper', 'sd-webui-check-tensors', 'sd-webui-color-enhance', 'sd-webui-cutoff',
                         'sd-webui-freeu', 'sd-webui-infinite-image-browsing', 'sd-webui-lora-block-weight', 'sd-webui-negpip',
                         'sd-webui-openpose-editor', 'sd-webui-pixelart', 'sd-webui-prompt-all-in-one', 'sd-webui-stablesr',
                         'sd-webui-supermerger', 'sdweb-merge-block-weighted-gui', 'sd_webui_SAG', 'stable-diffusion-prompt-pai',
                         'stable-diffusion-webui-anti-burn', 'stable-diffusion-webui-cafe-aesthetic', 'stable-diffusion-webui-model-toolkit',
                         'stable-diffusion-webui-pixelization', 'stable-diffusion-webui-two-shot', 'ultimate-upscale-for-automatic1111',
                         'Umi-AI']
07:06:27-041596 INFO     Verifying requirements
07:06:27-043096 WARNING  Setup complete with errors: 1 <-- (it's openOutpaint-webUI-extension, but it seems to be partially working)
07:06:27-045096 WARNING  See log file for more details: X:\AI\automatic\sdnext.log
07:06:27-057596 INFO     Command line args: []
07:06:33-257096 INFO     Load packages: {'torch': '2.3.0+cu121', 'diffusers': '0.28.0', 'gradio': '3.43.2'}
07:06:34-348096 INFO     VRAM: Detected=24.0 GB Optimization=none
07:06:34-353596 INFO     Engine: backend=Backend.DIFFUSERS compute=cuda device=cuda attention="Scaled-Dot-Product" mode=no_grad
07:06:34-539096 INFO     Device: device=NVIDIA GeForce RTX 3090 n=1 arch=sm_90 cap=(8, 6) cuda=12.1 cudnn=8801 driver=555.85
07:06:36-058597 INFO     Available VAEs: path="models\VAE" items=34
07:06:36-064097 INFO     Disabled extensions: ['sdnext-modernui', 'ABG_extension', 'SD-latent-mirroring', 'TokenMixer',
                         'multidiffusion-upscaler-for-automatic1111', 'sd-face-editor', 'sd-webui-additional-networks', 'sd-webui-animatediff',
                         'sd-webui-neutral-prompt', 'sdweb-merge-board', 'stable-diffusion-webui-Prompt_Generator',
                         'stable-diffusion-webui-aesthetic-gradients', 'stable-diffusion-webui-embedding-merge',
                         'stable-diffusion-webui-text2prompt', 'stable-diffusion-webui-visualize-cross-attention-extension', 'tagger',
                         'weight_gradient']
```

### Relevant log output

```shell
19:21:08-235462 INFO     Base: class=StableDiffusionPAGPipeline
Progress ?it/s                                              0% 0/40 00:00 ? Base
19:21:08-731963 ERROR    Processing: args={'prompt_embeds': tensor([[[-3.6035e-01,  3.4809e-04, -6.8298e-02,  ..., -4.4360e-01,
                                   -2.8638e-01,  4.7424e-02],
                                  [ 6.6895e-01, -6.3184e-01, -2.0581e-01,  ...,  5.6982e-01,
                                    1.7944e-01,  2.9004e-01],
                                  [ 1.0000e+00, -2.4902e-01, -1.7520e+00,  ...,  1.7520e+00,
                                   -3.1323e-01, -3.0029e-01],
                                  ...,
                                  [-7.0508e-01, -7.7344e-01, -3.1396e-01,  ..., -1.1895e+00,
                                   -7.6123e-01, -4.2822e-01],
                                  [-7.0312e-01, -7.6709e-01, -3.0420e-01,  ..., -1.1943e+00,
                                   -7.6270e-01, -4.1016e-01],
                                  [-6.6309e-01, -7.3389e-01, -2.6587e-01,  ..., -1.1562e+00,
                                   -7.5098e-01, -4.7607e-01]],

                                 [[-3.6035e-01,  3.4809e-04, -6.8298e-02,  ..., -4.4360e-01,
                                   -2.8638e-01,  4.7424e-02],
                                  [ 6.6895e-01, -6.3184e-01, -2.0581e-01,  ...,  5.6982e-01,
                                    1.7944e-01,  2.9004e-01],
                                  [ 1.0000e+00, -2.4902e-01, -1.7520e+00,  ...,  1.7520e+00,
                                   -3.1323e-01, -3.0029e-01],
                                  ...,
                                  [-7.0508e-01, -7.7344e-01, -3.1396e-01,  ..., -1.1895e+00,
                                   -7.6123e-01, -4.2822e-01],
                                  [-7.0312e-01, -7.6709e-01, -3.0420e-01,  ..., -1.1943e+00,
                                   -7.6270e-01, -4.1016e-01],
                                  [-6.6309e-01, -7.3389e-01, -2.6587e-01,  ..., -1.1562e+00,
                                   -7.5098e-01, -4.7607e-01]]], device='cuda:0', dtype=torch.float16), 'negative_prompt_embeds':
                         tensor([[[-3.6060e-01,  3.4809e-04, -6.8298e-02,  ..., -4.4385e-01,
                                   -2.8662e-01,  4.7424e-02],
                                  [-7.0508e-01,  4.9414e-01,  1.0996e+00,  ...,  2.8320e-01,
                                    2.0581e-01,  1.1865e+00],
                                  [-2.4280e-01, -2.7295e-01,  2.0254e+00,  ..., -3.1445e-01,
                                   -4.6436e-01,  1.8701e+00],
                                  ...,
                                  [ 4.5581e-01, -4.9438e-02, -5.9082e-01,  ..., -7.8467e-01,
                                   -3.7500e-01, -6.2927e-02],
                                  [ 4.4458e-01, -4.3854e-02, -5.9180e-01,  ..., -7.7881e-01,
                                   -3.8232e-01, -4.5013e-02],
                                  [ 4.7729e-01, -3.8422e-02, -5.4541e-01,  ..., -7.5537e-01,
                                   -3.4546e-01, -1.3049e-01]],

                                 [[-3.6060e-01,  3.4809e-04, -6.8298e-02,  ..., -4.4385e-01,
                                   -2.8662e-01,  4.7424e-02],
                                  [-7.0508e-01,  4.9414e-01,  1.0996e+00,  ...,  2.8320e-01,
                                    2.0581e-01,  1.1865e+00],
                                  [-2.4280e-01, -2.7295e-01,  2.0254e+00,  ..., -3.1445e-01,
                                   -4.6436e-01,  1.8701e+00],
                                  ...,
                                  [ 4.5581e-01, -4.9438e-02, -5.9082e-01,  ..., -7.8467e-01,
                                   -3.7500e-01, -6.2927e-02],
                                  [ 4.4458e-01, -4.3854e-02, -5.9180e-01,  ..., -7.7881e-01,
                                   -3.8232e-01, -4.5013e-02],
                                  [ 4.7729e-01, -3.8422e-02, -5.4541e-01,  ..., -7.5537e-01,
                                   -3.4546e-01, -1.3049e-01]]], device='cuda:0', dtype=torch.float16), 'guidance_scale': 6.5, 'generator':
                         [<torch._C.Generator object at 0x0000017C26426930>, <torch._C.Generator object at 0x0000017C26426A10>],
                         'callback_on_step_end': <function diffusers_callback at 0x0000017CDA80E560>, 'callback_on_step_end_tensor_inputs':
                         ['latents', 'prompt_embeds', 'negative_prompt_embeds'], 'num_inference_steps': 40, 'eta': 1.0, 'guidance_rescale': 0.6,
                         'output_type': 'latent', 'width': 448, 'height': 744, 'pag_scale': 3.5, 'pag_adaptive_scaling': 0.5,
                         'pag_applied_layers_index': ['m0']} shape '[6, 2, 46, 2, 28, 320]' is invalid for input of size 9999360
19:21:08-752463 ERROR    Processing: RuntimeError
╭────────────────────────────────────────────────────── Traceback (most recent call last) ──────────────────────────────────────────────────────╮
│ X:\AI\automatic\modules\processing_diffusers.py:118 in process_diffusers                                                                      │
│                                                                                                                                               │
│   117 │   │   #    base_args['image'] = set_latents(p)                                                                                        │
│ ❱ 118 │   │   output = shared.sd_model(**base_args) # pylint: disable=not-callable                                                            │
│   119 │   │   if isinstance(output, dict):                                                                                                    │
│                                                                                                                                               │
│ X:\AI\automatic\venv\lib\site-packages\torch\utils\_contextlib.py:115 in decorate_context                                                     │
│                                                                                                                                               │
│   114 │   │   with ctx_factory():                                                                                                             │
│ ❱ 115 │   │   │   return func(*args, **kwargs)                                                                                                │
│   116                                                                                                                                         │
│                                                                                                                                               │
│ X:\AI\automatic\modules\pag\pipe_sd.py:1354 in __call__                                                                                       │
│                                                                                                                                               │
│   1353 │   │   │   │   # predict the noise residual                                                                                           │
│ ❱ 1354 │   │   │   │   noise_pred = self.unet(                                                                                                │
│   1355 │   │   │   │   │   latent_model_input,                                                                                                │
│                                                                                                                                               │
│ X:\AI\automatic\venv\lib\site-packages\torch\nn\modules\module.py:1532 in _wrapped_call_impl                                                  │
│                                                                                                                                               │
│   1531 │   │   else:                                                                                                                          │
│ ❱ 1532 │   │   │   return self._call_impl(*args, **kwargs)                                                                                    │
│   1533                                                                                                                                        │
│                                                                                                                                               │
│ X:\AI\automatic\venv\lib\site-packages\torch\nn\modules\module.py:1582 in _call_impl                                                          │
│                                                                                                                                               │
│   1581 │   │   │                                                                                                                              │
│ ❱ 1582 │   │   │   result = forward_call(*args, **kwargs)                                                                                     │
│   1583 │   │   │   if _global_forward_hooks or self._forward_hooks:                                                                           │
│                                                                                                                                               │
│                                                            ... 6 frames hidden ...                                                            │
│                                                                                                                                               │
│ X:\AI\automatic\venv\lib\site-packages\diffusers\models\transformers\transformer_2d.py:448 in forward                                         │
│                                                                                                                                               │
│   447 │   │   │   else:                                                                                                                       │
│ ❱ 448 │   │   │   │   hidden_states = block(                                                                                                  │
│   449 │   │   │   │   │   hidden_states,                                                                                                      │
│                                                                                                                                               │
│ X:\AI\automatic\venv\lib\site-packages\torch\nn\modules\module.py:1532 in _wrapped_call_impl                                                  │
│                                                                                                                                               │
│   1531 │   │   else:                                                                                                                          │
│ ❱ 1532 │   │   │   return self._call_impl(*args, **kwargs)                                                                                    │
│   1533                                                                                                                                        │
│                                                                                                                                               │
│ X:\AI\automatic\venv\lib\site-packages\torch\nn\modules\module.py:1541 in _call_impl                                                          │
│                                                                                                                                               │
│   1540 │   │   │   │   or _global_forward_hooks or _global_forward_pre_hooks):                                                                │
│ ❱ 1541 │   │   │   return forward_call(*args, **kwargs)                                                                                       │
│   1542                                                                                                                                        │
│                                                                                                                                               │
│ X:\AI\automatic\modules\hidiffusion\hidiffusion.py:167 in forward                                                                             │
│                                                                                                                                               │
│   166 │   │   │   │   shift_size = (widow_size[0]//4*3, widow_size[1]//4*3)                                                                   │
│ ❱ 167 │   │   │   norm_hidden_states = window_partition(norm_hidden_states, widow_size, shift_size, H, W)                                     │
│   168                                                                                                                                         │
│                                                                                                                                               │
│ X:\AI\automatic\modules\hidiffusion\hidiffusion.py:112 in window_partition                                                                    │
│                                                                                                                                               │
│   111 │   │   │   │   │   │   x = torch.roll(x, shifts=(-shift_size, -shift_size), dims=(1, 2))                                               │
│ ❱ 112 │   │   │   │   x = x.view(B, H // window_size[0], window_size[0], W // window_size[1], window_size[1], C)                              │
│   113 │   │   │   │   windows = x.permute(0, 1, 3, 2, 4, 5).contiguous().view(-1, window_size[0], window_size[1], C)                          │
╰───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
RuntimeError: shape '[6, 2, 46, 2, 28, 320]' is invalid for input of size 9999360
```

### Backend

Diffusers

### Branch

Master

### Model

SD 1.5

### Acknowledgements

- [X] I have read the above and searched for existing issues
- [X] I confirm that this is classified correctly and its not an extension issue

**Comment by vladmandic** [comment]
a) hidiffusion is sensitive to non-square aspect ratios. try with standard resolutions and see if issue is gone or still present
b) you have over a dozen extensions. rule of a thumb is to disable extensions when troubleshooting, its impossible to predict how some things may interact. start sdnext with `--safe` flag so it temporarily disables extensions without you needing to do that for each manually.

**Comment by mart-hill** [comment]
Alrighty, I'll test it with clean run!

**Comment by mart-hill** [comment]
Confirmed, it happens with --safe flag, too:

First attempt:
![00000-20240605031447_3881769076_15b1d230-grid](https://github.com/vladmandic/automatic/assets/55754730/e43d099e-0d38-40d0-bef9-857ddae94fee)

Second attempt:
![00001-20240605031529_3881769076_29917329-grid](https://github.com/vladmandic/automatic/assets/55754730/c0d2a309-a7c5-486a-aa12-2de87cc4d1e0)

**Comment by vladmandic** [comment]
original backend doesn't have hidifussion at all. this is due to hidifussion error. did you check with square aspect ratio?

**Comment by mart-hill** [comment]
The funny thing is this happens when I have HiDiffusion unchecked. Also, the option remains visible for the original backend.

**Comment by vladmandic** [comment]
> Also, the option remains visible for the original backend.

it doesn't do anything. i can hide it for original backend.

> The funny thing is this happens when I have HiDiffusion OFF

once the problem occurs, its hard to fully disable hidiffusion without reloading the model, it gets its hooks deep.
does the problem occur on freshly started server with hidiffusion never being turned on? that would be bad.
does the problem occur on freshly started server with hidiffusion on and using square image aspect ratio?

**Comment by mart-hill** [comment]
1. Ahh. Makes sense then it's this 'influential'.
2. Yes.
3. Yes.

![00039-20240605173115_3881769102_364d2558-grid](https://github.com/vladmandic/automatic/assets/55754730/09c0b84e-54da-45b7-bc00-bad940e0e8db)
![00040-20240605173343_3881769102_1d7dcd67-grid](https://github.com/vladmandic/automatic/assets/55754730/196f728d-fd24-4a40-a336-a623f3bf5b70)




There's a curiosity - if I generate the second, supposedly 'right' image, and then just remove the word from the prompt, and restore it back (without generating the image in-between), to the letter, the image will be generated the 'right' way, as if the text encoder cached the result.

**Comment by vladmandic** [comment]
ok, now we're getting somewhere! yes, text encoder results are cached unless prompt changes. i didn't expect hidiffusion on/off should reset cache, but i can easily add that.

**Comment by mart-hill** [comment]
Would that be that text encoder cache should also be reset on fresh 'boot'/restart the UI, too?

**Comment by vladmandic** [comment]
i just pushed an update to `dev` branch, can you try?
lets keep aspect ratio as square for now, best to deal with one problem at the time.

**Comment by mart-hill** [comment]
OK, let's try! 🙂 
I hope it'll go well, cause somehow, now, with original backend, I have `01:18:10-075253 WARNING  PAG: pipeline=<class 'ldm.models.diffusion.ddpm.LatentDiffusion'> required=StableDiffusionPipeline` warning every generation, and, just before that, last diffusers generations of mine were suddenly total 'colored glass' garbage. Testing! 🙂

**Comment by mart-hill** [comment]
BTW, HiDiffusion option ON won't allow me to generate a landscape 960x532 image (x2 in 2nd pass)
```
02:10:26-042754 ERROR    Processing: args={'prompt_embeds': tensor([[[-0.3811,  0.0188, -0.0574,  ..., -0.4910, -0.3035,  0.0645],
                                  [ 0.0339,  0.0684,  0.0251,  ..., -1.2510,  1.0176, -0.0444],
                                  [ 1.1436,  0.5557, -0.0476,  ..., -1.7881,  0.6406,  0.6924],
                                  ...,
                                  [-0.5479,  1.4131, -0.3916,  ..., -1.0576,  0.0410,  0.5596],
                                  [-0.5410,  1.4375, -0.3394,  ..., -1.0898,  0.0488,  0.5747],
                                  [-0.5400,  1.4980, -0.4563,  ..., -1.1318,  0.0664,  0.5781]]],
                                device='cuda:0', dtype=torch.float16), 'negative_prompt_embeds': tensor([[[-0.3811,  0.0188, -0.0574,  ...,
                         -0.4910, -0.3035,  0.0645],
                                  [-0.8628,  0.1738,  1.4531,  ...,  0.5698,  0.2063,  1.1602],
                                  [-0.3188, -0.0186,  2.0312,  ..., -0.8135, -0.4753,  1.1553],
                                  ...,
                                  [ 0.6289, -0.4014, -0.2942,  ...,  1.4424, -1.5674,  0.0735],
                                  [ 0.6270, -0.3833, -0.2769,  ...,  1.4648, -1.5811,  0.0863],
                                  [ 0.6387, -0.3462, -0.2522,  ...,  1.4385, -1.5078,  0.0494]]],
                                device='cuda:0', dtype=torch.float16), 'guidance_scale': 7.5, 'generator': [<torch._C.Generator object at
                         0x000001C62D361990>], 'callback_on_step_end': <function diffusers_callback at 0x000001C60BABD090>,
                         'callback_on_step_end_tensor_inputs': ['latents', 'prompt_embeds', 'negative_prompt_embeds'],
                         'num_inference_steps': 40, 'eta': 1.0, 'guidance_rescale': 0.7, 'output_type': 'latent', 'width': 960, 'height':
                         536} shape '[2, 2, 33, 2, 60, 320]' is invalid for input of size 5145600
02:10:26-053755 ERROR    Processing: RuntimeError

Blah, blah

RuntimeError: shape '[2, 2, 33, 2, 60, 320]' is invalid for input of size 5145600
```


...but will allow reversed-AR one. 🙂

**Comment by vladmandic** [comment]
OK, this is something deeper, I need to spend more time on this.

Btw, disabling embeds cache has around 0.2-0.6s impact, depending on several items, not even close to 15s you mentioned, unless there is some weird offloading happening - are you running with lowvram or something like that?

**Comment by mart-hill** [comment]
That's my settings on the 'compute' section, cause I can't use anything about 'compile' on Windows, right? 
![image](https://github.com/vladmandic/automatic/assets/55754730/76aa1bd3-89d7-43f3-83d1-a112b3c7331c)

DIffusers settings:
![image](https://github.com/vladmandic/automatic/assets/55754730/3fd59bdb-58d7-4aab-a190-2796b9f50bb7)

Inference settings:
![image](https://github.com/vladmandic/automatic/assets/55754730/c56280fd-d0b5-4499-ba34-7965bd630e4d)

**Comment by vladmandic** [comment]
nothing there stands out as obviously wrong.

also, we need to split issues:
- hidiffusion failures on some non-square aspect ratio
- difference in geneate output between first run and second run
- delay on start of generate

for first, lets table it for now.
for second, i need to spend time on it trying to reproduce and analyze.
for third, can you create a separate issue and upload debug log - but please set `SD_PROCESS_DEBUG=true` env variable before starting sdnext so it captures more info

**Comment by vladmandic** [comment]
ok, i've reproduced the problem of 1st vs all subsequent generations and its specific to batch-size=2, none of the other parameters matter.
i hope to have a solution soon.

unrelated: your prompt parser is set to compel and compel does not understand `()` syntax. you've probably set compel as a mistake?
unrelated: your negative prompt is very bad:
- its heavily unbalanced: e.g. "physically-uncommon" at strength 1.3 and "extra digits" at strength 1.54 are just going to burn things.
- use of `|` has no meaning here at all. you cannot alternate between words in a negative prompt.
- i have no idea what [bad : wrong : morbid]" is even supposed to do and i wrote prompt parser :)

**Comment by mart-hill** [comment]
> nothing there stands out as obviously wrong.
> 
> also, we need to split issues:
> 
> * hidiffusion failures on some non-square aspect ratio
> * difference in geneate output between first run and second run
> * delay on start of generate
> 
> for first, lets table it for now. for second, i need to spend time on it trying to reproduce and analyze. but that remains the key issue here. for third, can you create a separate issue and upload debug log - but please set `SD_PROCESS_DEBUG=true` env variable before starting sdnext so it captures more info

Thank you for your hard work, the third issue seems to be gone! The processing with HiDIffusion ON starts almost immediately now. The bad news is the images sometimes still seems to be different between first run and second run... Gulp. 🙂 

With 2nd pass and forced HiresFix:
![00021-20240607070243_3984525576_001aeb78-grid](https://github.com/vladmandic/automatic/assets/55754730/82442455-0a66-45b7-9d73-08826224cada)
![00022-20240607070522_3984525576_211f8444-grid](https://github.com/vladmandic/automatic/assets/55754730/d5e3417e-6dc4-44a5-8c33-1570e45e12ff)


It's hard to catch this bug! Maybe it happens just after model loading? First pair was also rendered twice just after RevAnimated 2.0 model was loaded.

I've loaded a RevAnimated 1.2.2 model and immediately after I generated the two pairs of the images, one after another. HiDiffusion os ON, of course. Drat. :)
![00042-20240607074329_3984525576_edf529c2-grid](https://github.com/vladmandic/automatic/assets/55754730/55b20061-e07f-4249-98b1-5f5680e250e2)
![00043-20240607074726_3984525576_6b8c8818-grid](https://github.com/vladmandic/automatic/assets/55754730/034d7e6f-0407-4aa1-89f4-b4283e28a00f)



I'm terribly sorry about the negative prompt, it is sort of 'brainless', indeed. I thought `|` or `:` would be processed by `dynamic-prompt` extension though. Latest dev updates (I'm on 5574833f commit) made it complaining a bit though (`AttributeError: module 'pyparsing' has no attribute 'Opt'`). :)

**Comment by mart-hill** [comment]
I just tested HiDiffusion with a model, which creates practically mutants at 768x768 pixels, and realized, that the FIRST image just after loading seems to be the 'right' one (no mutants anymore, HiDiffusion works its magic), but second generation is actually an image without HiDiffusion (total mutant mess). Would that explain something?

**Comment by vladmandic** [comment]
first part - general torch

ah, this turned out to be a deep rathole, but i think i figured it out - and its something that will explain few things overall.

you're using both pag and hidiffusion - and each adds its own layer of complexity, up to a point where it hides the underlying issue.

underlying issue is that torch by default is tiny bit undeterministic, even if using precise seed.
to be precise, its almost-guaranteed deterministic for single stream, but when you're using batch-size=2, that means two streams are running concurrently and tiny nanosecond differences can cause non-deterministic results. you can reproduce this by running same generate with batch-size=1 vs batch-size=2 and see few pixels change. not many, just few, but enough to cascade up into pag and then into hidiffusion where small differences become big ones.

i've added settings -> compute -> use deterministic
i don't want to enable it by default since visual impact for most users is barely visible and it does come with performance penalty

**Comment by vladmandic** [comment]
second part - hidiffusion

as most of "research" work, authors come up with something cool, but there is no way to easily turn it on/off, so it becomes a patchwork - its just not designed to ever be disabled once its enabled.
a while ago i did a patch that fixes this for sdxl, but forgot about sd15, additional code is now in that should address this.
this caused hidiffusion to be applied on first generate, but not on subsequent.

additionally, hidiffusion itself is based on introducing a tiny bit of randomness.
you can disable randomness by disabling settings -> inference -> hiduffusion -> msw-msa, but that reduces effectiveness of hidiffusion by a lot.

**Comment by mart-hill** [comment]
> first part - general torch
> 
> ah, this turned out to be a deep rathole, but i think i figured it out - and its something that will explain few things overall.
> 
> you're using both pag and hidiffusion - and each adds its own layer of complexity, up to a point where it hides the underlying issue.
> 
> underlying issue is that torch by default is tiny bit undeterministic, even if using precise seed. to be precise, its almost-guaranteed deterministic for single stream, but when you're using batch-size=2, that means two streams are running concurrently and tiny nanosecond differences can cause non-deterministic results. you can reproduce this by running same generate with batch-size=1 vs batch-size=2 and see few pixels change. not many, just few, but enough to cascade up into pag and then into hidiffusion where small differences become big ones.
> 
> i've added settings -> compute -> use deterministic i don't want to enable it by default since visual impact for most users is barely visible and it does come with performance penalty

I though PAG is enabled by default, I only started using diffusers backend very recently, after that 'jumbo' update for the `master` branch. Should I disable it?
Thank you for the detailed explanations! I'll update the `dev` branch and check. :)

**Comment by vladmandic** [comment]
PAG gets enabled if your attention strength is not zero, default is zero.
no issues with it being enabled, just added one more variable into the mix so complicates troubleshooting.

For wiki, most of those explanations I write on-the-spot, I'd really appreciate if community (you?) took them and did a writeup?

**Comment by mart-hill** [comment]
I'd be using your words if I did that, it wouldn't be right to do so, right?

**The issue is gone for good - thank you!** :)

**Comment by vladmandic** [comment]
yeah, but it would get done - I can either make stuff or document stuff, not both 😂
